backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/ui/web/order/BOMExploderCommand.java

```diff
@@ -130,17 +139,59 @@ public BOMExploderCommand(@Nullable final BOMUse bomToUse,
 			final OrderLineCandidate lineCandidate = initialCandidate.toBuilder()
 					.productId(bomLineProductId)
 					.attributes(attributes)
-					.qty(bomLineQty)
+					.qty(Quantitys.create(bomLineQty, bomUomId))
 					.compensationGroupId(compensationGroupId)
-					.uomId(bomUomId)
 					.explodedFromBOMLineId(bomLineId)
 					.build();
 			result.addAll(this.toBuilder()
 					.initialCandidate(lineCandidate)
 					.build()
-					.execute());  // recurse
+					.execute0());  // recurse
 		}
 
 		return result;
 	}
+
+	private List<OrderLineCandidate> reduceCandidates(final List<OrderLineCandidate> result)
+	{
+		final ArrayList<OrderLineCandidate> reduced = new ArrayList<>();
+		for (final OrderLineCandidate candidate : result)
+		{
+			if (candidate.getQty().signum() == 0)
+			{
+				continue;
+			}
+			if (candidate.getQty().signum() < 0)
+			{
+				throw new AdempiereException("Invalid qty: " + candidate.getQty());
+			}
+			if (candidate.getQty().toBigDecimal().compareTo(BigDecimal.ZERO) == 0)
+			{
+				continue;
+			}
+
+			final Optional<OrderLineCandidate> orderLineCandidate = reduced.stream()
+					.filter(cand -> cand.getProductId().equals(candidate.getProductId()))
+					.findFirst();
+			if (orderLineCandidate.isPresent())
+			{
+				final OrderLineCandidate olCandidate = orderLineCandidate.get();
+				if (!olCandidate.getAttributes().equals(candidate.getAttributes()))
+				{
+					throw new AdempiereException("Cannot explode BOM because attributes are different: " + olCandidate.getAttributes() + " and " + candidate.getAttributes());
+				}
+				reduced.remove(olCandidate);
```

Comment:
i always get uncomfortable when i see the remove method used on large objects... relying entirely on the equals implementation.


Instead i would rewrite the whole method as follows:
```java
		result.stream()
				.collect(ImmutableMap.toImmutableMap(
						OrderLineCandidate::getProductId,
						olCand -> olCand,
						(olCand1, olCand2) -> combine(olCand1, olCand2)));
```
....
```java
	private static OrderLineCandidate combine(@NonNull OrderLineCandidate olCand1, @NonNull OrderLineCandidate olCand2)
	{
		///...
	}
```

Reply:
Yes, I like your approach better. Was not thinking about it since in my early draft I used more fields to qualify orderLinCeandidates for reduction. Fixed.